### DISCUSSION TITLE
Rollbacks are not supported for DDL statements.

### DISCUSSION THREAD

**Discussion by jokineno** [Snowflake-Labs/schemachange#372] [Q&A]
We use the schemachange library to run database migrations, and I would like to highlight a limitation: rollbacks are not supported for DDL operations like CREATE TABLE or ALTER TABLE statements because Snowflake executes these operations in their own separate transactions.

A possible workaround is to take a backup of resources using Snowflake's built-in CLONE feature (e.g., CREATE TABLE  CLONE ) before running any migrations (SQL files containing one or multiple statements). If schemachange deploy returns an exit code 1, a rollback can be triggered by SWAPping the resources, such as tables, and then removing the backups. This approach would effectively revert the changes in case of any issues during the migration.

Anyone who has multiple role grants in the database resources should pay attention how those are impacted after CLONE/SWAP.

I believe adding a similar rollback mechanism as a wrapper around the deploy command could make the schemachange library more flexible and attractive to users. It would provide a much-needed safety net for DDL migrations, aligning with best practices for robust database management.

Also, I am happy to hear how others have solved rollback for DDL operations.

**Comment by sfc-gh-tmathew** [comment]
Thanks for the detailed suggestion! Rollback support for DDL operations is definitely a known limitation.

This topic has come up in the community before, and we acknowledge it's a common expectation when working with Snowflake's DDL statements. Your CLONE/SWAP approach is one of the workarounds that some teams use. Our advice has been to change forward instead of rollback as workaround.

## Current Status

Rollback/revert capabilities are on our radar and listed in our backlog of future ideas, but we haven't scheduled implementation yet due to:

1. **Complexity** - Proper rollback support for DDL requires significant design work
2. **Current workarounds** - Teams can implement their own strategies using CLONE/SWAP, careful migration organization, alter in lower environment, OR build rollback scripts along with deploy scripts and run it as separate calls.
3. **Resource constraints** - We're focusing on other priorities in near-term releases

## Recommended Approaches Today

Until native rollback support is added, here are the patterns we recommend:

- **Test thoroughly in lower environments** before production
- **Organize scripts by dependency** - independent objects first, dependent objects in order
- **Use `--dry-run` flag** to preview changes
- **Implement custom CLONE/SWAP wrapper** if your use case requires it (as you described)
- **Use separate deployments** by passing different `--root-folder` values for each dependency layer

If others have implemented rollback patterns or have specific use cases, please share them here - community input helps us prioritize features!

**Comment by sfc-gh-tmathew** [comment]
Adding reference to past PRs and Issues opened on this topic. 
https://github.com/Snowflake-Labs/schemachange/issues/19

Rollback functionality will potentially be a separate sub-command and figure out a workflow that can be supported with rollback scenarios. 

What workarounds have others tried with any success?
